"레스토랑 조회 구현하기" 제출

anpaul0615 opened the pull request.

anpaul0615 committed:
프로젝트 이전

anpaul0615 committed:
App 컴포넌트 내 입력 관련 코드 제거

anpaul0615 committed:
RestaurantCategoriesContainer 컴포넌트 테스트 작성 및 구현

- categories 관련 표출 구현
- categories 관련 상태관리 구현
- restaurants 관련 <ol /> 와 categories 관련 <ol /> 을 id 로 구분 및 테스트 내 식별방식 변경

anpaul0615 commented:
생각보다 처리할게 많아서 완료하지 못한 상태로 제출했습니다...
조금 늦어지겠지만 그래도 꼭 완료해놓을게요..!

hannut91 commented on src/components/CategoryButtonList.jsx:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+
+export default function CategoryButtonList({ categories, handleClickCategory }) {
```
카테고리 목록이 중요하지 버튼인지는 중요하지 않은 것 같아요

anpaul0615 replied:
그냥 카테고리목록이 아니라, 버튼으로 클릭할수 있는 카테고리 목록 아닌가요?

![image](https://user-images.githubusercontent.com/8488507/86580433-a8332300-bfb9-11ea-8d29-b02a7bbeaab0.png)

anpaul0615 replied:
카테고리도 버튼도 목록도 관심사대로 분리해야할것같습니다.

이 컴포넌트가 카테고리에 종속될 필요는 없어보이고, 기존의 `List` 와는 구분하는게 좋아보입니다.
컴포넌트 명칭을 `ButtonList` 정도로 수정하면 될것같습니다.

hannut91 commented on src/components/InputBox.test.jsx:
```diff
@@ -0,0 +1,31 @@
+
+  const renderComponent = ({ inputProperty, inputValue }) => render((
+    <InputBox
+      inputProperty={inputProperty}
```
이미 컴포넌트에 Input이란 말이 있으니 name으로 바로 받아도 될 것 같아요

anpaul0615 replied:
네 알겠습니다..!

hannut91 commented on src/stores/stores.test.js:
```diff
@@ -0,0 +1,7 @@
+import RestaurantStore from './stores';
```
왜 레스토랑 스토어인지 모르겠어요. 스토어 입장에서는 레스토랑이 중요하지 않을거에요. 그냥 store로만 해줘도 괜찮을 것 같아요

anpaul0615 replied:
사실 그게 맞는말같아요... redux 는 스토어가 하나니깐요.

anpaul0615 replied:
> 스토어 입장에서는 레스토랑이 중요하지 않을거에요. 

이말에 전적으로 동의해요.
이 스토어가 어떤 스토어인지 `stores.test.js` 에서 확인할필요는 없어보여요.

anpaul0615 replied:
그러면 또 궁금한게 있는데요.

redux 는 단일스토어잖아요? 그런데 왜 소스코드 파일명을 stores 라고 하나요?

hannut91 replied:
저는 `store.js`가 맞는 것 같습니다.

anpaul0615 replied:
그렇죠? 저도 그렇게 생각해요..!

anpaul0615 replied:
> 그런데 왜 소스코드 파일명을 stores 라고 하나요?

이 댓글을 적을때는 강의에서 `stores.js` 라고 했다고 인식하고있었는데요.
다시 확인해보니깐 제가 혼자서 착각한거였어요.. 

제대로 수정해놓을게요..!

hannut91 commented on src/stores/reducer.test.js:
```diff
@@ -0,0 +1,120 @@
+  context('when action-type is invalid', () => {
+    it('do nothing', () => {
+      const previousState = {
+        someStatePropperty: Math.random() * 10,
```
이거는 왜 사용되는지 잘 모르겠어요

anpaul0615 replied:
invalid 한 액션타입이 들어왔을때, 아무일도 일어나지 않는다는 것을 표현한겁니다.
(= previousState 에 어떤 속성들이 있든지간에, nextState 는 previousState 와 항상 동일함)

hannut91 replied:
someStateProperty가 없어도 동작하지 않나요?

anpaul0615 replied:
맞아요 없어도 동작해요.
그런데도 굳이 특정 속성을 명시한건 nextState 가 previousState 와 동일하다는걸 표현하기위해서에요.

anpaul0615 replied:
다시생각해보니깐, `someStatePropperty: Math.random() * 10` 보다는  `someState: 'someValue'` 가 더 잘읽힐것같네요..!

hannut91 commented on src/stores/reducer.test.js:
```diff
@@ -0,0 +1,120 @@
+    const previousState = {
+      categories: [],
+    };
+    const categories = CATEGORIES;
```
이 줄은 없애고 바로 CATEGORIES를 써도 될 것 같아요

anpaul0615 replied:
이것도 많이 고민했었는데요.
위 아래 다른 테스트 코드들의 모양? 을 맞춰주려고 일부러 저렇게 작성한거거든요.

테스트코드 읽을때 모양을 맞춰주면 쭉쭉쭉 읽히면서 다른부분만 눈에 들어오는데, 갑자기 코드 모양이 달라지게되면, 어 이건 어디서가져온거지? 하면서 다시 위쪽 선언부를 찾아본다는거죠. 그게 저는 굉장히 불편하더라구요.

사실 따지고보면 이런것도 다 중복이긴한데.. 그래도 코드리딩을 좀더 편하게하려고 코드모양을 맞춰봤습니다.
 
이부분은 어떻게 생각하세요?

hannut91 replied:
어디를 말씀하시는 건지 잘 모르겠어요. `categories`를 사용하는 곳은 한 곳 밖에 보이지 않아요

anpaul0615 replied:
말씀하신대로 `categories` 를 사용하는 곳은 한곳밖에없네요... ㅎㅎㅎ

비슷한 모양의 코드로는
34번라인 : `const categories = CATEGORIES;`
50번라인 : `const category = CATEGORIES[0];`
가 있네요.. ㅎㅎ

위 댓글에서 언급했었던 코드모양은 
```
// Given
const previousState = {  ...  };  // reducer 이전 상태값
const ~~~ = ... ;  // reducer, action 으로 전달하기 위한 값
```
으로 어떤 정형화된 모양? 을 만들어보려고 했었던 것 같습니다.

hannut91 commented on src/stores/reducer.js:
```diff
@@ -0,0 +1,78 @@
+
+  [UPDATE_RESTAURANT_PROPERTY]: (previousState, payload) => {
+    const { restaurant } = previousState;
+    const { propertyName, propertyValue } = payload;
```
property라는 단어가 코드를 이해하는데 도움이 되지않는 것 같아요

anpaul0615 replied:
정~~말정말 궁금한데요. 저는 `name`, `value` 로만 작성하는게 더 이해가 안되거든요... 

그냥 딸랑 `name`, `value` 만 써놓으면 코드 읽다가 이게 무슨 name 인지 무슨 value 인지 헷갈릴때가 있더라구요.
물론 지금 과제코드는 몇줄안되서 크게 안느껴지겠지만요..

그래서 제생각에는 이런 추상적인 변수명은 명시적으로 정확하게 작성해줘야할것같거든요..
이 name 이란 값은 어디어디의 name 이다. 이 value 란 값은 어디어디의 value 이다 라고요.

어떻게 생각하세요?

anpaul0615 replied:
혹시라도 `propertyName`, `propertyValue` 에서 `property` 가 중복으로 보여서 불편한거라면,
`property: { name, value }` 쓰는거까지는 동의할수있을것같아요.

anpaul0615 replied:
그리고 또 사실은
`[UPDATE_RESTAURANT_PROPERTY]: (previousState, payload) => {}` 
이 코드 자체가 레스토랑의 name, category, address 입력박스 이벤트를 하나로 처리하기위해 저렇게 `propertyName` 으로 구분하면서 비슷한 함수들을 합쳐버린건데요.

그렇다면 name 입력박스로 부터 들어온 값을 `propertyName`, `propertyValue` 으로 구분하지않고 그냥 `name` 과 `value` 로 구분하는게 더 보기 헷갈린다고 생각하거든요.

name 입력박스로 부터 들어온 name 과 value ..? 
저는 코드에서 이건 어디의 name 인거지? 라는 물음표가 생겨서 엄청 헷갈렸었거든요.

hannut91 replied:
변수의 사용범위가 좁을수록 간결한 이름을 사용하게 됩니다. propertyValue와 propertyName이 이 함수 안에서만 사용되기 때문에 저는 name과 value만 가지고도 문맥상 파악할 수 있다고 생각했습니다. 이름은 정답이 없기 때문에 고민해보면서 더 좋은 이름을 발견하면 되고 또 개발을 하다 보면 더 좋은 이름이 생기기도하고, 적합한 이름을 찾았다고 생각했는데 기능이 변경되어 아에 쓸모없는 이름이 될 수도 있습니다.

anpaul0615 replied:
아 그렇군요. 
"함수 안에서만 사용되기 때문에 ..문맥상 파악할 수 있다" 라는 말이 확 와닿네요.
앞으로 변수이름 정할때 이 변수가 어떤범위에서 사용되는지도 반영해보겠습니다..!

hannut91 commented on src/containers/RestaurantInputContainer.test.jsx:
```diff
@@ -0,0 +1,65 @@
+    const { getAllByRole } = renderComponent();
+    getAllByRole('textbox').forEach((inputBox) => {
+      const infoProperty = inputBox.name;
+      const infoValue = restaurant[infoProperty];
```
info라는 말이 아무런 의미를 가지고 있지 않아요

anpaul0615 replied:
앗 이건 지난번 과제에서 고친다고 고쳤는데.. 
코드옮겨오면서 뭔가 실수가 있었던것같네요. 고쳐놓겠습니다.

hannut91 commented on src/containers/RestaurantInputContainer.jsx:
```diff
@@ -0,0 +1,34 @@
+
+  const dispatch = useDispatch();
+
+  function handleupdateRestaurantProperty(ev) {
```
간단하게 사용하실거면 e 한 글자만 쓰던지 아니면 event를 전체 다 써주는게 좋을 것 같아요. ev로 줄여쓰는 걸 못본 것 같아요

anpaul0615 replied:
그렇게 볼수도있을것같아요.

개인적으로 생각했을때 소스코드에서의 `e` 는 `error`, `exception`, `event` 등 엄청나게 많은 중요 키워드들의 접두어라서 조금 구분하는게 좋겠다고 생각하거든요.

참고할게요.

anpaul0615 replied:
그리고 아래는 ev 로 줄여쓰는 코드들이에요.
예전부터 종종 써왔었던거라 저는 ev 를 엄청 많이 봤었거든요.

https://github.com/mobxjs/mobx/search?q=ev&unscoped_q=ev
https://github.com/microsoft/TypeScript/search?q=ev&unscoped_q=ev

한번 확인해보세요.

hannut91 replied:
그렇군요 저도 처음봤습니다 ㅎㅎ 감사합니다

anpaul0615 committed:
~/services/api 테스트 추가

anpaul0615 committed:
~/action/action-creators 비동기 액션함수 오타 수정

anpaul0615 committed:
컴포넌트 명칭 및 변수 명칭 수정

- <RestaurantCategoryContainer /> → <RestaurantCategoryContainer />
- <CategoryButtonList /> → <ButtonList />
- <InputBox /> : inputProperty/inputValue → property/value
- <List /> : restaurants/categories → items

anpaul0615 committed:
불필요한 App 컴포넌트 테스트 제거

anpaul0615 committed:
코드 구조 변경

- action/* 제거 → actions/category, actions/restaurant 분리 구현
- stores/* 제거 → store, reducer 분리 구현
- sevices/api → sevices/category 분리 구현

anpaul0615 committed:
actions/category 테스트 작성

anpaul0615 commented:
가이드해주신대로 코드를 전반적으로 고쳐봤어요.
특히 각 컴포넌트별 관심사대로 최대한 나눠보려고 노력해봤는데.. 방향이 맞는지 궁금하네요..ㅎㅎ

기능구현을 다 못해서 e2e 테스트는 통과하지못했구요.
현재까지 구현한 범위까지만 해서 커버리지 100 맞췄어요.

많은 코멘트 부탁드릴게요..!

anpaul0615 committed:
services 관련 코드 디렉토리 변경

- services/* → services/api/*

anpaul0615 commented:
아래 저장소 위키 참고하면서 계속 지저분한 코드 정리하고 있습니다..ㅎㅎ

https://github.com/diegohaz/arc/wiki/Containers
https://github.com/diegohaz/arc/wiki/API-service
https://github.com/diegohaz/arc/wiki/Redux-modules

anpaul0615 committed:
current-category-id 를 seleced-category-id 로 변경

anpaul0615 committed:
ButtonList 컴포넌트 selectedId 속성 기능 구현

anpaul0615 committed:
actions/region 테스트 작성 및 구현

anpaul0615 committed:
region 관련 reducer 테스트 작성 및 구현

anpaul0615 committed:
services/region 테스트 작성 및 구현

anpaul0615 committed:
ButtonList 컴포넌트 selectedId 관련 표출오류 수정

anpaul0615 committed:
RestaurantRegionContainer 컴포넌트 테스트 작성 및 구현

anpaul0615 committed:
부적절한 컴포넌트 명칭 수정

- RestaurantCategoryButtonContainer → RestaurantCategoryContainer
- RestaurantRegionButtonContainer → RestaurantRegionContainer

anpaul0615 committed:
RestaurantRegionContainer 를 App 컴포넌트에 적용

anpaul0615 committed:
List, ButtonList 컴포넌트 내 ol 엘리먼트를 ul 로 변경

anpaul0615 committed:
RestaurantInputContainer 제거

anpaul0615 committed:
services/region.test 오타 수정

anpaul0615 committed:
reducer 에서 RestaurantInputContainer 관련 코드 제거

anpaul0615 committed:
reducer.test 오타 수정

anpaul0615 committed:
services/restaurant 테스트 작성 및 구현

anpaul0615 committed:
기존 actions/restaurants 삭제

anpaul0615 committed:
actions/restaurant 테스트 작성 및 구현

anpaul0615 committed:
restaurant 관련 reducer 테스트 작성 및 구현

anpaul0615 committed:
RestaurantDisplayContainer 컴포넌트 리팩토링

- selectedCategoryId, selectedRegionId 를 바라보는 useEffect 구현

anpaul0615 committed:
loadCategories, loadRegions 관심사를 App 컴포넌트로 이동

anpaul0615 committed:
e2e 테스트 동작 오류 수정

- loadRestaurants 에서 selectedCategoryId, selectedRegionId 누락 예외상황 처리
- ButtonList 컴포넌트에서 " (V)" 띄어쓰기 제거
- List 컴포넌트에서 명칭만 표출하도록 변경

hannut91 reviewed (approved):
고생하셨습니다 ㅎㅎ 이번 과제를 통해 무엇을 배우셨나요?

anpaul0615 commented:
이번과제에서는 관심사의 분리에 대해서 많은 고민을 해볼수있었습니다.

Presentational 컴포넌트와 Container 컴포넌트 사이의 결합도를 느슨하게 하기위한 여러가지 시도를 해봤습니다. 
컴포넌트 명칭 수정, Presentational 컴포넌트의 props 변수명 수정, redux 관련 코드 디렉토리 구조 변경 등등 여러가지를 시도해보고 고쳐보면서 이번과제에서의 학습포인트를 찾아보려고 노력했습니다.

이번 과제에서의 시행착오를 떠올리면서 다음과제부터는 더 좋은 코드를 작성해보겠습니다.

p.s
늦은시간에도 최선을 다해 제 질문에 대답해주신 윤석님(@hannut91) 께 진심으로 진심으로 감사인사 드립니다..!
잘 이해가 안되는 모습을 답답해하시면서도 그래도 어떻게든 알려주려고하는 그 모든 시간들이 저에게 정말 많은 도움이 됐습니다. 때지난 과제였지만, 그래도 발벗고 도와주셔서 감사했습니다.
